staking/tests/simple_test.rs

```diff
@@ -0,0 +1,366 @@
+        USERS[3],
+        FTAction::Transfer {
+            from: USERS[3].into(),
+            to: USERS[0].into(),
+            amount: 100000,
+        },
+    );
+    assert!(!res.main_failed());
+
+    let res = rw_token.send(USERS[3], FTAction::BalanceOf(USERS[0].into()));
+    assert!(res.contains(&(USERS[3], FTEvent::Balance(100000).encode())));
+}
+
+/// Calculates the reward produced so far
+fn produced(staking: &mut Staking, time: u64) -> u128 {
+    staking.reward_produced
+        + (time - staking.produced_time) as u128 / staking.distribution_time as u128
+}
+
+/// Updates the reward produced so far and calculates tokens per stake
+fn update_reward(staking: &mut Staking, time: u64) {
+    let reward_produced_at_now = produced(staking, time);
+
+    if reward_produced_at_now > staking.reward_produced {
+        let produced_new = reward_produced_at_now - staking.reward_produced;
+
+        if staking.total_staked > 0 {
+            staking.tokens_per_stake = staking
+                .tokens_per_stake
+                .saturating_add(produced_new * (1 << 20) / staking.total_staked);
+        }
+
+        staking.reward_produced = staking.reward_produced.saturating_add(produced_new);
+    }
+}
+
+/// Calculates the reward of the staker that is currently avaiable
+fn calc_reward(staking: &Staking, source: &ActorId) -> u128 {
+    if let Some(staker) = staking.stakers.get(source) {
+        return (staker.balance * staking.tokens_per_stake) / (1 << 20) + staker.reward_allowed
+            - staker.reward_debt
+            - staker.distributed;
+    }
+
+    panic!("calc_reward(): Staker {:?} not found", source);
+}
+
+#[test]
+fn stake() {
+    let sys = System::new();
+    init_staking(&sys);
+    init_staking_token(&sys);
+    init_reward_token(&sys);
+    sys.init_logger();
+    let staking = sys.get_program(1);
+
+    let res = staking.send(USERS[4], StakingAction::Stake(1000));
+    assert!(res.contains(&(USERS[4], StakingEvent::StakeAccepted(1000).encode())));
+
+    let res = staking.send(USERS[5], StakingAction::Stake(3000));
+    assert!(res.contains(&(USERS[5], StakingEvent::StakeAccepted(3000).encode())));
+}
+
+#[test]
+fn send_reward() {
+    let sys = System::new();
+
+    init_staking(&sys);
+    init_staking_token(&sys);
+    init_reward_token(&sys);
+    sys.init_logger();
+    let st = sys.get_program(1);
+
+    let time = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .expect("Time went backwards")
+        .as_secs();
+
+    let mut staking = Staking {
+        distribution_time: 10000,
+        produced_time: time,
+        ..Default::default()
+    };
+
+    let res = st.send(USERS[4], StakingAction::Stake(1500));
+    assert!(res.contains(&(USERS[4], StakingEvent::StakeAccepted(1500).encode())));
+
+    update_reward(&mut staking, time);
+    staking.stakers.insert(
+        USERS[4].into(),
+        Staker {
+            reward_debt: (1500 * staking.tokens_per_stake) / (1 << 20),

```

Comment:
Can we use the same constant and approach as in the program itself?

Reply:
Fixed